// @remotion/lambda/dist/api/iam-validation/simulate-rule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.simulateRule = void 0;
const client_iam_1 = require("@aws-sdk/client-iam");
const lambda_client_1 = require("@remotion/lambda-client");
const simulateRule = async (options) => {
    var _a;
    try {
        const res = await lambda_client_1.LambdaClientInternals.getIamClient(options.region, options.requestHandler).send(new client_iam_1.SimulatePrincipalPolicyCommand({
            ActionNames: options.actionNames,
            PolicySourceArn: options.arn,
            ResourceArns: options.resource,
        }));
        return ((_a = res.EvaluationResults) !== null && _a !== void 0 ? _a : []).map((pol) => {
            return {
                decision: pol.EvalDecision,
                name: pol.EvalActionName,
            };
        });
    }
    catch (err) {
        // Sometimes the AWS Rate limit hits. In that case we retry up to two times
        // after waiting for 2 seconds.
        if (options.retries <= 0) {
            throw err;
        }
        await new Promise((resolve) => {
            setTimeout(resolve, 2000);
        });
        return (0, exports.simulateRule)({ ...options, retries: options.retries - 1 });
    }
};
exports.simulateRule = simulateRule;
