// @remotion/licensing/dist/esm/index.mjs
// src/register-usage-point.ts
var HOST = "https://www.remotion.pro";
var registerUsageEvent = async ({
  apiKey,
  host,
  succeeded,
  event
}) => {
  const res = await fetch(`${HOST}/api/track/register-usage-point`, {
    method: "POST",
    body: JSON.stringify({
      event,
      apiKey,
      host,
      succeeded
    }),
    headers: {
      "Content-Type": "application/json"
    }
  });
  const json = await res.json();
  if (json.success) {
    return {
      billable: json.billable,
      classification: json.classification
    };
  }
  if (!res.ok) {
    throw new Error(json.error);
  }
  const read = await res.json();
  return read;
};
// src/get-usage.ts
var getUsage = async ({
  apiKey,
  since
}) => {
  const res = await fetch(`${HOST}/api/track/get-usage`, {
    method: "POST",
    body: JSON.stringify({
      apiKey,
      since: since ?? null
    }),
    headers: {
      "Content-Type": "application/json"
    }
  });
  const json = await res.json();
  if (json.success) {
    return {
      cloudRenders: json.cloudRenders,
      webcodecConversions: json.webcodecConversions
    };
  }
  throw new Error(json.error);
};
export {
  registerUsageEvent,
  getUsage
};
