// @remotion/lambda-client/dist/cjs/constants.js
var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __moduleCache = /* @__PURE__ */ new WeakMap;
var __toCommonJS = (from) => {
  var entry = __moduleCache.get(from), desc;
  if (entry)
    return entry;
  entry = __defProp({}, "__esModule", { value: true });
  if (from && typeof from === "object" || typeof from === "function")
    __getOwnPropNames(from).map((key) => !__hasOwnProp.call(entry, key) && __defProp(entry, key, {
      get: () => from[key],
      enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
    }));
  __moduleCache.set(from, entry);
  return entry;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {
      get: all[name],
      enumerable: true,
      configurable: true,
      set: (newValue) => all[name] = () => newValue
    });
};

// src/constants.ts
var exports_constants = {};
__export(exports_constants, {
  getSitesKey: () => getSitesKey,
  RENDER_FN_PREFIX: () => RENDER_FN_PREFIX,
  REMOTION_BUCKET_PREFIX: () => REMOTION_BUCKET_PREFIX,
  MIN_TIMEOUT: () => MIN_TIMEOUT,
  MIN_MEMORY: () => MIN_MEMORY,
  MIN_EPHEMERAL_STORAGE_IN_MB: () => MIN_EPHEMERAL_STORAGE_IN_MB,
  MAX_TIMEOUT: () => MAX_TIMEOUT,
  MAX_MEMORY: () => MAX_MEMORY,
  MAX_EPHEMERAL_STORAGE_IN_MB: () => MAX_EPHEMERAL_STORAGE_IN_MB,
  LOG_GROUP_PREFIX: () => LOG_GROUP_PREFIX,
  LAMBDA_INSIGHTS_PREFIX: () => LAMBDA_INSIGHTS_PREFIX,
  LAMBDA_CONCURRENCY_LIMIT_QUOTA: () => LAMBDA_CONCURRENCY_LIMIT_QUOTA,
  DEFAULT_TIMEOUT: () => DEFAULT_TIMEOUT,
  DEFAULT_REGION: () => DEFAULT_REGION,
  DEFAULT_OUTPUT_PRIVACY: () => DEFAULT_OUTPUT_PRIVACY,
  DEFAULT_MEMORY_SIZE: () => DEFAULT_MEMORY_SIZE,
  DEFAULT_MAX_RETRIES: () => DEFAULT_MAX_RETRIES,
  DEFAULT_FRAMES_PER_LAMBDA: () => DEFAULT_FRAMES_PER_LAMBDA,
  DEFAULT_EPHEMERAL_STORAGE_IN_MB: () => DEFAULT_EPHEMERAL_STORAGE_IN_MB,
  DEFAULT_CLOUDWATCH_RETENTION_PERIOD: () => DEFAULT_CLOUDWATCH_RETENTION_PERIOD,
  BINARY_NAME: () => BINARY_NAME
});
module.exports = __toCommonJS(exports_constants);
var MIN_MEMORY = 512;
var MAX_MEMORY = 10240;
var DEFAULT_MEMORY_SIZE = 2048;
var DEFAULT_TIMEOUT = 120;
var MIN_TIMEOUT = 15;
var MAX_TIMEOUT = 900;
var DEFAULT_FRAMES_PER_LAMBDA = 20;
var BINARY_NAME = "remotion lambda";
var DEFAULT_REGION = "us-east-1";
var DEFAULT_MAX_RETRIES = 1;
var MAX_EPHEMERAL_STORAGE_IN_MB = 10240;
var DEFAULT_EPHEMERAL_STORAGE_IN_MB = 2048;
var MIN_EPHEMERAL_STORAGE_IN_MB = 512;
var DEFAULT_OUTPUT_PRIVACY = "public";
var DEFAULT_CLOUDWATCH_RETENTION_PERIOD = 14;
var RENDER_FN_PREFIX = "remotion-render-";
var LOG_GROUP_PREFIX = "/aws/lambda/";
var LAMBDA_INSIGHTS_PREFIX = "/aws/lambda-insights";
var getSitesKey = (siteId) => `sites/${siteId}`;
var LAMBDA_CONCURRENCY_LIMIT_QUOTA = "L-B99A9384";
var REMOTION_BUCKET_PREFIX = "remotionlambda-";
