// @remotion/lambda/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateWebhookSignature = exports.simulatePermissions = exports.renderVideoOnLambda = exports.renderStillOnLambda = exports.renderMediaOnLambda = exports.presignUrl = exports.LambdaInternals = exports.getUserPolicy = exports.getSites = exports.getRolePolicy = exports.getRenderProgress = exports.getRegions = exports.getOrCreateBucket = exports.getFunctions = exports.getFunctionInfo = exports.getCompositionsOnLambda = exports.getAwsClient = exports.estimatePrice = exports.downloadMedia = exports.deploySite = exports.deployFunction = exports.deleteSite = exports.deleteRender = exports.deleteFunction = void 0;
const lambda_client_1 = require("@remotion/lambda-client");
Object.defineProperty(exports, "deleteFunction", { enumerable: true, get: function () { return lambda_client_1.deleteFunction; } });
Object.defineProperty(exports, "deleteRender", { enumerable: true, get: function () { return lambda_client_1.deleteRender; } });
Object.defineProperty(exports, "estimatePrice", { enumerable: true, get: function () { return lambda_client_1.estimatePrice; } });
Object.defineProperty(exports, "getAwsClient", { enumerable: true, get: function () { return lambda_client_1.getAwsClient; } });
Object.defineProperty(exports, "getCompositionsOnLambda", { enumerable: true, get: function () { return lambda_client_1.getCompositionsOnLambda; } });
Object.defineProperty(exports, "getFunctions", { enumerable: true, get: function () { return lambda_client_1.getFunctions; } });
Object.defineProperty(exports, "renderVideoOnLambda", { enumerable: true, get: function () { return lambda_client_1.renderVideoOnLambda; } });
Object.defineProperty(exports, "validateWebhookSignature", { enumerable: true, get: function () { return lambda_client_1.validateWebhookSignature; } });
const no_react_1 = require("remotion/no-react");
const delete_site_1 = require("./api/delete-site");
Object.defineProperty(exports, "deleteSite", { enumerable: true, get: function () { return delete_site_1.deleteSite; } });
const deploy_function_1 = require("./api/deploy-function");
Object.defineProperty(exports, "deployFunction", { enumerable: true, get: function () { return deploy_function_1.deployFunction; } });
const deploy_site_1 = require("./api/deploy-site");
Object.defineProperty(exports, "deploySite", { enumerable: true, get: function () { return deploy_site_1.deploySite; } });
const download_media_1 = require("./api/download-media");
Object.defineProperty(exports, "downloadMedia", { enumerable: true, get: function () { return download_media_1.downloadMedia; } });
const get_function_info_1 = require("./api/get-function-info");
Object.defineProperty(exports, "getFunctionInfo", { enumerable: true, get: function () { return get_function_info_1.getFunctionInfo; } });
const get_or_create_bucket_1 = require("./api/get-or-create-bucket");
Object.defineProperty(exports, "getOrCreateBucket", { enumerable: true, get: function () { return get_or_create_bucket_1.getOrCreateBucket; } });
const get_regions_1 = require("./api/get-regions");
Object.defineProperty(exports, "getRegions", { enumerable: true, get: function () { return get_regions_1.getRegions; } });
const simulate_1 = require("./api/iam-validation/simulate");
Object.defineProperty(exports, "simulatePermissions", { enumerable: true, get: function () { return simulate_1.simulatePermissions; } });
const suggested_policy_1 = require("./api/iam-validation/suggested-policy");
Object.defineProperty(exports, "getRolePolicy", { enumerable: true, get: function () { return suggested_policy_1.getRolePolicy; } });
Object.defineProperty(exports, "getUserPolicy", { enumerable: true, get: function () { return suggested_policy_1.getUserPolicy; } });
const internals_1 = require("./internals");
Object.defineProperty(exports, "LambdaInternals", { enumerable: true, get: function () { return internals_1.LambdaInternals; } });
/**
 * @deprecated Import this from `@remotion/lambda-client` instead
 */
const renderMediaOnLambda = no_react_1.NoReactInternals.ENABLE_V5_BREAKING_CHANGES
    ? () => {
        throw new Error('renderMediaOnLambda() has moved to `@remotion/lambda-client`. Please import it from there.');
    }
    : lambda_client_1.renderMediaOnLambda;
exports.renderMediaOnLambda = renderMediaOnLambda;
/**
 * @deprecated Import this from `@remotion/lambda-client` instead
 */
const getRenderProgress = no_react_1.NoReactInternals.ENABLE_V5_BREAKING_CHANGES
    ? () => {
        throw new Error('getRenderProgress() has moved to `@remotion/lambda-client`. Please import it from there.');
    }
    : lambda_client_1.getRenderProgress;
exports.getRenderProgress = getRenderProgress;
/**
 * @deprecated Import this from `@remotion/lambda-client` instead
 */
const renderStillOnLambda = no_react_1.NoReactInternals.ENABLE_V5_BREAKING_CHANGES
    ? () => {
        throw new Error('renderStillOnLambda() has moved to `@remotion/lambda-client`. Please import it from there.');
    }
    : lambda_client_1.renderStillOnLambda;
exports.renderStillOnLambda = renderStillOnLambda;
/**
 * @deprecated Import this from `@remotion/lambda-client` instead
 */
const presignUrl = no_react_1.NoReactInternals.ENABLE_V5_BREAKING_CHANGES
    ? () => {
        throw new Error('presignUrl() has moved to `@remotion/lambda-client`. Please import it from there.');
    }
    : lambda_client_1.presignUrl;
exports.presignUrl = presignUrl;
/**
 * @deprecated Import this from `@remotion/lambda-client` instead
 */
const getSites = no_react_1.NoReactInternals.ENABLE_V5_BREAKING_CHANGES
    ? () => {
        throw new Error('getSites() has moved to `@remotion/lambda-client`. Please import it from there.');
    }
    : lambda_client_1.getSites;
exports.getSites = getSites;
