// @remotion/lambda-client/dist/aws-clients.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStsClient = exports.getServiceQuotasClient = exports.getIamClient = exports.getLambdaClient = exports.getCloudWatchLogsClient = void 0;
const get_service_client_1 = require("./get-service-client");
const getCloudWatchLogsClient = (region, requestHandler) => {
    return (0, get_service_client_1.getServiceClient)({
        region,
        service: 'cloudwatch',
        customCredentials: null,
        forcePathStyle: false,
        requestHandler,
    });
};
exports.getCloudWatchLogsClient = getCloudWatchLogsClient;
const getLambdaClient = (region, 
// eslint-disable-next-line @typescript-eslint/no-unused-vars
_timeoutInTest, requestHandler) => {
    return (0, get_service_client_1.getServiceClient)({
        region,
        service: 'lambda',
        customCredentials: null,
        forcePathStyle: false,
        requestHandler,
    });
};
exports.getLambdaClient = getLambdaClient;
const getIamClient = (region, requestHandler) => {
    return (0, get_service_client_1.getServiceClient)({
        region,
        service: 'iam',
        customCredentials: null,
        forcePathStyle: false,
        requestHandler,
    });
};
exports.getIamClient = getIamClient;
const getServiceQuotasClient = (region, requestHandler) => {
    return (0, get_service_client_1.getServiceClient)({
        region,
        service: 'servicequotas',
        customCredentials: null,
        forcePathStyle: false,
        requestHandler,
    });
};
exports.getServiceQuotasClient = getServiceQuotasClient;
const getStsClient = (region, requestHandler) => {
    return (0, get_service_client_1.getServiceClient)({
        region,
        service: 'sts',
        customCredentials: null,
        forcePathStyle: false,
        requestHandler,
    });
};
exports.getStsClient = getStsClient;
