// @remotion/lambda-client/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.REMOTION_BUCKET_PREFIX = exports.LAMBDA_CONCURRENCY_LIMIT_QUOTA = exports.getSitesKey = exports.LAMBDA_INSIGHTS_PREFIX = exports.LOG_GROUP_PREFIX = exports.RENDER_FN_PREFIX = exports.DEFAULT_CLOUDWATCH_RETENTION_PERIOD = exports.DEFAULT_OUTPUT_PRIVACY = exports.MIN_EPHEMERAL_STORAGE_IN_MB = exports.DEFAULT_EPHEMERAL_STORAGE_IN_MB = exports.MAX_EPHEMERAL_STORAGE_IN_MB = exports.DEFAULT_MAX_RETRIES = exports.DEFAULT_REGION = exports.BINARY_NAME = exports.DEFAULT_FRAMES_PER_LAMBDA = exports.MAX_TIMEOUT = exports.MIN_TIMEOUT = exports.DEFAULT_TIMEOUT = exports.DEFAULT_MEMORY_SIZE = exports.MAX_MEMORY = exports.MIN_MEMORY = void 0;
exports.MIN_MEMORY = 512;
exports.MAX_MEMORY = 10240;
exports.DEFAULT_MEMORY_SIZE = 2048;
exports.DEFAULT_TIMEOUT = 120;
exports.MIN_TIMEOUT = 15;
exports.MAX_TIMEOUT = 900;
exports.DEFAULT_FRAMES_PER_LAMBDA = 20;
exports.BINARY_NAME = 'remotion lambda';
exports.DEFAULT_REGION = 'us-east-1';
exports.DEFAULT_MAX_RETRIES = 1;
exports.MAX_EPHEMERAL_STORAGE_IN_MB = 10240;
// TODO: In V5, Enable set this to 10240
exports.DEFAULT_EPHEMERAL_STORAGE_IN_MB = 2048;
exports.MIN_EPHEMERAL_STORAGE_IN_MB = 512;
exports.DEFAULT_OUTPUT_PRIVACY = 'public';
exports.DEFAULT_CLOUDWATCH_RETENTION_PERIOD = 14;
exports.RENDER_FN_PREFIX = 'remotion-render-';
exports.LOG_GROUP_PREFIX = '/aws/lambda/';
exports.LAMBDA_INSIGHTS_PREFIX = '/aws/lambda-insights';
const getSitesKey = (siteId) => `sites/${siteId}`;
exports.getSitesKey = getSitesKey;
exports.LAMBDA_CONCURRENCY_LIMIT_QUOTA = 'L-B99A9384';
exports.REMOTION_BUCKET_PREFIX = 'remotionlambda-';
