// @remotion/lambda-client/dist/get-aws-client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAwsClient = void 0;
const CloudWatchSDK = __importStar(require("@aws-sdk/client-cloudwatch-logs"));
const IamSdk = __importStar(require("@aws-sdk/client-iam"));
const LambdaSDK = __importStar(require("@aws-sdk/client-lambda"));
const S3SDK = __importStar(require("@aws-sdk/client-s3"));
const ServiceQuotasSDK = __importStar(require("@aws-sdk/client-service-quotas"));
const StsSdk = __importStar(require("@aws-sdk/client-sts"));
const get_service_client_1 = require("./get-service-client");
/*
 * @description Exposes full access to the AWS SDK used by Remotion, allowing interaction with AWS infrastructure beyond provided functionalities.
 * @see [Documentation](https://remotion.dev/docs/lambda/getawsclient)
 */
const getAwsClient = ({ region, service, customCredentials, forcePathStyle, requestHandler, }) => {
    return {
        client: (0, get_service_client_1.getServiceClient)({
            region,
            service,
            customCredentials: customCredentials !== null && customCredentials !== void 0 ? customCredentials : null,
            forcePathStyle: forcePathStyle !== null && forcePathStyle !== void 0 ? forcePathStyle : false,
            requestHandler,
        }),
        sdk: {
            lambda: LambdaSDK,
            cloudwatch: CloudWatchSDK,
            iam: IamSdk,
            s3: S3SDK,
            servicequotas: ServiceQuotasSDK,
            sts: StsSdk,
        }[service],
    };
};
exports.getAwsClient = getAwsClient;
