// @remotion/lambda-client/dist/get-sites.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSites = exports.internalGetSites = void 0;
const aws_provider_1 = require("./aws-provider");
const constants_1 = require("./constants");
const make_s3_url_1 = require("./make-s3-url");
const internalGetSites = async ({ region, forceBucketName, providerSpecifics, forcePathStyle, requestHandler, }) => {
    var _a;
    const remotionBuckets = forceBucketName
        ? await providerSpecifics.getBuckets({
            region,
            forceBucketName,
            forcePathStyle,
            requestHandler,
        })
        : await providerSpecifics.getBuckets({
            region,
            forceBucketName: null,
            forcePathStyle,
            requestHandler,
        });
    const accountId = await providerSpecifics.getAccountId({ region });
    const sites = {};
    for (const bucket of remotionBuckets) {
        const ls = await providerSpecifics.listObjects({
            bucketName: bucket.name,
            prefix: (0, constants_1.getSitesKey)(''),
            region,
            expectedBucketOwner: accountId,
            forcePathStyle,
            requestHandler,
        });
        for (const file of ls) {
            const siteKeyMatch = (_a = file.Key) === null || _a === void 0 ? void 0 : _a.match(/sites\/([0-9a-zA-Z-!_.*'()]+)\/(.*)$/);
            if (!siteKeyMatch) {
                throw new Error(`A file was found in the bucket "${bucket.name}" with the key ${file.Key} which is an unexpected folder structure. Delete this file.`);
            }
            const [, siteId] = siteKeyMatch;
            if (!sites[siteId]) {
                sites[siteId] = {
                    sizeInBytes: 0,
                    bucketName: bucket.name,
                    lastModified: null,
                    id: siteId,
                    serveUrl: (0, make_s3_url_1.makeS3ServeUrl)({
                        bucketName: bucket.name,
                        region,
                        subFolder: (0, constants_1.getSitesKey)(siteId),
                    }),
                };
            }
            if (file.LastModified) {
                const currentLastModified = sites[siteId].lastModified;
                if (currentLastModified === null ||
                    file.LastModified.getTime() > currentLastModified) {
                    sites[siteId].lastModified = file.LastModified.getTime();
                }
            }
            if (file.Size) {
                sites[siteId].sizeInBytes += file.Size;
            }
        }
    }
    const sitesArray = Object.keys(sites).map((siteId) => {
        return sites[siteId];
    });
    return { sites: sitesArray, buckets: remotionBuckets };
};
exports.internalGetSites = internalGetSites;
/*
 * @description Gets an array of Remotion projects in Cloud Storage, in your GCP project.
 * @see [Documentation](https://remotion.dev/docs/cloudrun/getsites)
 */
const getSites = ({ region, forceBucketName, forcePathStyle, requestHandler, }) => {
    return (0, exports.internalGetSites)({
        region,
        forceBucketName: forceBucketName !== null && forceBucketName !== void 0 ? forceBucketName : null,
        forcePathStyle: forcePathStyle !== null && forcePathStyle !== void 0 ? forcePathStyle : false,
        providerSpecifics: aws_provider_1.awsImplementation,
        requestHandler: requestHandler !== null && requestHandler !== void 0 ? requestHandler : null,
    });
};
exports.getSites = getSites;
