// @remotion/lambda/dist/api/get-or-create-bucket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOrCreateBucket = void 0;
const lambda_client_1 = require("@remotion/lambda-client");
const serverless_1 = require("@remotion/serverless");
/*
 * @description Creates a Cloud Storage bucket for Remotion Cloud Run in your GCP project. If one already exists, it will get returned instead.
 * @see [Documentation](https://remotion.dev/docs/cloudrun/getorcreatebucket)
 */
const getOrCreateBucket = (options) => {
    var _a, _b, _c;
    return (0, serverless_1.internalGetOrCreateBucket)({
        region: options.region,
        enableFolderExpiry: (_a = options.enableFolderExpiry) !== null && _a !== void 0 ? _a : null,
        customCredentials: (_b = options.customCredentials) !== null && _b !== void 0 ? _b : null,
        providerSpecifics: lambda_client_1.LambdaClientInternals.awsImplementation,
        forcePathStyle: false,
        skipPutAcl: false,
        requestHandler: (_c = options.requestHandler) !== null && _c !== void 0 ? _c : null,
    });
};
exports.getOrCreateBucket = getOrCreateBucket;
