// @remotion/lambda/dist/api/get-function-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFunctionInfo = void 0;
const client_lambda_1 = require("@aws-sdk/client-lambda");
const lambda_client_1 = require("@remotion/lambda-client");
const constants_1 = require("@remotion/lambda-client/constants");
/*
 * @description Gets information about a function given its name and region.
 * @see [Documentation](https://remotion.dev/docs/lambda/getfunctioninfo)
 */
const getFunctionInfo = async ({ region, functionName, logLevel, requestHandler, }) => {
    var _a, _b, _c, _d, _e;
    lambda_client_1.LambdaClientInternals.validateAwsRegion(region);
    const [functionInfo, version] = await Promise.all([
        lambda_client_1.LambdaClientInternals.getLambdaClient(region, undefined, requestHandler).send(new client_lambda_1.GetFunctionCommand({
            FunctionName: functionName,
        })),
        (0, lambda_client_1.getFunctionVersion)({
            functionName,
            region,
            logLevel: logLevel !== null && logLevel !== void 0 ? logLevel : 'info',
            requestHandler,
        }),
    ]);
    return {
        functionName,
        timeoutInSeconds: (_a = functionInfo.Configuration) === null || _a === void 0 ? void 0 : _a.Timeout,
        memorySizeInMb: (_b = functionInfo.Configuration) === null || _b === void 0 ? void 0 : _b.MemorySize,
        version,
        diskSizeInMb: (_e = (_d = (_c = functionInfo.Configuration) === null || _c === void 0 ? void 0 : _c.EphemeralStorage) === null || _d === void 0 ? void 0 : _d.Size) !== null && _e !== void 0 ? _e : constants_1.DEFAULT_EPHEMERAL_STORAGE_IN_MB,
    };
};
exports.getFunctionInfo = getFunctionInfo;
