// @remotion/shapes/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeTriangle = exports.makeStar = exports.makeRect = exports.makePolygon = exports.makePie = exports.makeHeart = exports.makeEllipse = exports.makeCircle = exports.Triangle = exports.Star = exports.Rect = exports.Polygon = exports.Pie = exports.Heart = exports.Ellipse = exports.Circle = void 0;
var circle_1 = require("./components/circle");
Object.defineProperty(exports, "Circle", { enumerable: true, get: function () { return circle_1.Circle; } });
var ellipse_1 = require("./components/ellipse");
Object.defineProperty(exports, "Ellipse", { enumerable: true, get: function () { return ellipse_1.Ellipse; } });
var heart_1 = require("./components/heart");
Object.defineProperty(exports, "Heart", { enumerable: true, get: function () { return heart_1.Heart; } });
var pie_1 = require("./components/pie");
Object.defineProperty(exports, "Pie", { enumerable: true, get: function () { return pie_1.Pie; } });
var polygon_1 = require("./components/polygon");
Object.defineProperty(exports, "Polygon", { enumerable: true, get: function () { return polygon_1.Polygon; } });
var rect_1 = require("./components/rect");
Object.defineProperty(exports, "Rect", { enumerable: true, get: function () { return rect_1.Rect; } });
var star_1 = require("./components/star");
Object.defineProperty(exports, "Star", { enumerable: true, get: function () { return star_1.Star; } });
var triangle_1 = require("./components/triangle");
Object.defineProperty(exports, "Triangle", { enumerable: true, get: function () { return triangle_1.Triangle; } });
var make_circle_1 = require("./utils/make-circle");
Object.defineProperty(exports, "makeCircle", { enumerable: true, get: function () { return make_circle_1.makeCircle; } });
var make_ellipse_1 = require("./utils/make-ellipse");
Object.defineProperty(exports, "makeEllipse", { enumerable: true, get: function () { return make_ellipse_1.makeEllipse; } });
var make_heart_1 = require("./utils/make-heart");
Object.defineProperty(exports, "makeHeart", { enumerable: true, get: function () { return make_heart_1.makeHeart; } });
var make_pie_1 = require("./utils/make-pie");
Object.defineProperty(exports, "makePie", { enumerable: true, get: function () { return make_pie_1.makePie; } });
var make_polygon_1 = require("./utils/make-polygon");
Object.defineProperty(exports, "makePolygon", { enumerable: true, get: function () { return make_polygon_1.makePolygon; } });
var make_rect_1 = require("./utils/make-rect");
Object.defineProperty(exports, "makeRect", { enumerable: true, get: function () { return make_rect_1.makeRect; } });
var make_star_1 = require("./utils/make-star");
Object.defineProperty(exports, "makeStar", { enumerable: true, get: function () { return make_star_1.makeStar; } });
var make_triangle_1 = require("./utils/make-triangle");
Object.defineProperty(exports, "makeTriangle", { enumerable: true, get: function () { return make_triangle_1.makeTriangle; } });
