// @remotion/lambda/dist/cli/commands/sites/rmall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sitesRmallSubcommand = exports.SITES_RMALL_COMMAND = void 0;
const cli_1 = require("@remotion/cli");
const lambda_client_1 = require("@remotion/lambda-client");
const serverless_1 = require("@remotion/serverless");
const delete_site_1 = require("../../../api/delete-site");
const args_1 = require("../../args");
const get_aws_region_1 = require("../../get-aws-region");
const confirm_1 = require("../../helpers/confirm");
const log_1 = require("../../log");
exports.SITES_RMALL_COMMAND = 'rmall';
const sitesRmallSubcommand = async (logLevel, implementation) => {
    var _a;
    const region = (0, get_aws_region_1.getAwsRegion)();
    const deployedSites = await (0, lambda_client_1.getSites)({
        region,
    });
    const bucketName = (_a = args_1.parsedLambdaCli['force-bucket-name']) !== null && _a !== void 0 ? _a : (await (0, serverless_1.internalGetOrCreateBucket)({
        region,
        enableFolderExpiry: false,
        customCredentials: null,
        providerSpecifics: implementation,
        forcePathStyle: false,
        skipPutAcl: false,
        requestHandler: null,
    })).bucketName;
    for (const site of deployedSites.sites) {
        if (!(await (0, confirm_1.confirmCli)({
            delMessage: `Site ${site.id} in bucket ${site.bucketName} (${cli_1.CliInternals.formatBytes(site.sizeInBytes)}): Delete? (Y/n)`,
            allowForceFlag: true,
        }))) {
            continue;
        }
        const { totalSizeInBytes: totalSize } = await (0, delete_site_1.deleteSite)({
            bucketName,
            siteName: site.id,
            region,
            onAfterItemDeleted: ({ itemName }) => {
                log_1.Log.info({ indent: false, logLevel }, cli_1.CliInternals.chalk.gray(`Deleted ${itemName}`));
            },
        });
        log_1.Log.info({ indent: false, logLevel }, `Deleted site ${site.id} and freed up ${cli_1.CliInternals.formatBytes(totalSize)}.`);
    }
};
exports.sitesRmallSubcommand = sitesRmallSubcommand;
