// @remotion/lambda/dist/functions/helpers/read-with-progress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lambdaDownloadFileWithProgress = void 0;
const client_s3_1 = require("@aws-sdk/client-s3");
const s3_request_presigner_1 = require("@aws-sdk/s3-request-presigner");
const lambda_client_1 = require("@remotion/lambda-client");
const renderer_1 = require("@remotion/renderer");
const lambdaDownloadFileWithProgress = async ({ bucketName, key, region, expectedBucketOwner, outputPath, onProgress, customCredentials, logLevel, forcePathStyle, requestHandler, }) => {
    const client = lambda_client_1.LambdaClientInternals.getS3Client({
        region,
        customCredentials,
        forcePathStyle,
        requestHandler,
    });
    const command = new client_s3_1.GetObjectCommand({
        Bucket: bucketName,
        ExpectedBucketOwner: customCredentials ? undefined : expectedBucketOwner,
        Key: key,
    });
    const presigned = await (0, s3_request_presigner_1.getSignedUrl)(client, command);
    const { to, sizeInBytes } = await renderer_1.RenderInternals.downloadFile({
        url: presigned,
        onProgress: ({ downloaded, percent, totalSize }) => {
            // On Lambda, it should always be a number
            onProgress({
                downloaded,
                percent: percent,
                totalSize: totalSize,
            });
        },
        to: () => outputPath,
        indent: false,
        logLevel,
    });
    return { sizeInBytes, to };
};
exports.lambdaDownloadFileWithProgress = lambdaDownloadFileWithProgress;
